### PR TITLE
Job: Respect ControllerRef

### PR DESCRIPTION
**Pull request by enisoc**
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings Job into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that Job does not fight with other controllers over control of Pods.

Ref: #24433

**Special notes for your reviewer**:

**Release note**:

```release-note
Job controller now respects ControllerRef to avoid fighting over Pods.
```
cc @erictune @kubernetes/sig-apps-pr-reviews

### CODE REVIEW

File: pkg/controller/job/jobcontroller_test.go [hunk 1]
```diff
@@ -274,6 +275,22 @@ func TestControllerSyncJob(t *testing.T) {
 		if int32(len(fakePodControl.DeletePodName)) != tc.expectedDeletions {
 			t.Errorf("%s: unexpected number of deletes.  Expected %d, saw %d\n", name, tc.expectedDeletions, len(fakePodControl.DeletePodName))
 		}
+		// Each create should have an accompanying ControllerRef.
+		if len(fakePodControl.ControllerRefs) != int(tc.expectedCreations) {
+			t.Errorf("%s: unexpected number of ControllerRefs.  Expected %d, saw %d\n", name, tc.expectedCreations, len(fakePodControl.ControllerRefs))
+		}
+		// Make sure the ControllerRefs are correct.
+		for _, controllerRef := range fakePodControl.ControllerRefs {
+			if got, want := controllerRef.APIVersion, "batch/v1"; got != want {
```
Comment (soltysh) [inline pkg/controller/job/jobcontroller_test.go]: Can you verify the entire ControllerRef object, iow. including the Name and UID?

**Review by soltysh — APPROVED** [review]
A nit, but mostly LGTM.

**Comment by kow3ns** [comment]
Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[pkg/controller/job/jobcontroller.go, line 51 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176#-Ke0b5o3VVM_ffDQ6uux:-Ke0b5o3VVM_ffDQ6uuy:b-rdoodb) ([raw file](https://github.com/kubernetes/kubernetes/blob/38c4f5d95158e9040482d95c9c7558ee9461382c/pkg/controller/job/jobcontroller.go#L51)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = batch.SchemeGroupVersion.WithKind("Job")
> ```

Probably should be package private

---

*[pkg/controller/job/jobcontroller.go, line 149 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176#-Ke0bHk_nlteKT_tzpyf:-Ke0bHk_nlteKT_tzpyg:b-95lgwe) ([raw file](https://github.com/kubernetes/kubernetes/blob/38c4f5d95158e9040482d95c9c7558ee9461382c/pkg/controller/job/jobcontroller.go#L149)):*
> ```Go
> func (jm *JobController) getPodJobs(pod *v1.Pod) []*batch.Job {
> 	jobs, err := jm.jobLister.GetPodJobs(pod)
> 	if err != nil {
> ```

This is already here, but you have been removing th

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42176)*

**Comment by enisoc** [comment]
@k8s-bot non-cri e2e test this

**Comment by enisoc** [comment]
@k8s-bot cvm gce e2e test this

**Comment by enisoc** [comment]
@k8s-bot non-cri e2e test this

**Comment by enisoc** [comment]
@soltysh Please take a look. I believe all comments are addressed.

**Comment by soltysh** [comment]
@ethernetdan @calebamiles is this ok to merge for 1.6 release?

**Comment by caesarxuchao** [comment]
/approve

Approve the changes to client.

**Comment by lavalamp** [comment]
/approve

(registry changes)

**Comment by soltysh** [comment]
@lavalamp since head is open for 1.7 I think we can drop do-not-merge, right?

**Comment by soltysh** [comment]
@k8s-bot test this

**Comment by liggitt** [comment]
have the appropriate fixes been made to avoid racing the garbage controller readopting a pod on an orphaning deletion of a job?

**Comment by fejta** [comment]
/assign @lavalamp 

See https://github.com/kubernetes/kubernetes/pull/42176#issuecomment-289751249

**Comment by enisoc** [comment]
I've removed LGTM because I have some changes planned to bring this up to date with the other ControllerRef PRs that moved forward during the 1.6 cycle.

**Comment by enisoc** [comment]
This is ready for final re-review.

I've added 3 commits to fix the equivalent of https://github.com/kubernetes/kubernetes/pull/42938 and https://github.com/kubernetes/kubernetes/pull/43034. I made no changes to the previous commits other than rebasing on master.

File: pkg/controller/job/jobcontroller_test.go [hunk 2]
```diff
@@ -598,7 +599,8 @@ func TestJobPodLookup(t *testing.T) {
 	}
 	for _, tc := range testCases {
 		sharedInformerFactory.Batch().V1().Jobs().Informer().GetIndexer().Add(tc.job)
-		if job := manager.getPodJob(tc.pod); job != nil {
+		if jobs := manager.getPodJobs(tc.pod); len(jobs) > 0 {
+			job := jobs[0]
```
Comment (janetkuo) [inline pkg/controller/job/jobcontroller_test.go]: Also make sure `len(jobs)==1`

**Review by janetkuo — COMMENTED** [review]
LGTM in general

File: pkg/controller/job/jobcontroller.go [hunk 3]
```diff
@@ -167,14 +171,32 @@ func (jm *JobController) addPod(obj interface{}) {
...
+			// It's controlled by a different type of controller.
+			return
+		}
+		job, err := jm.jobLister.Jobs(pod.Namespace).Get(controllerRef.Name)
+		if err != nil {
+			return
+		}
 		jobKey, err := controller.KeyFunc(job)
 		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("Couldn't get key for job %#v: %v", job, err))
```
Comment (enisoc) [inline pkg/controller/job/jobcontroller.go]: In a comment on another ControllerRef PR, the reviewer asked me to remove this type of logging during watch handlers.

File: pkg/controller/job/jobcontroller_test.go (hunk 2 above)
Comment (enisoc) [inline pkg/controller/job/jobcontroller_test.go]: Done.

**Review by janetkuo — APPROVED** [review]
LGTM

**Pull request merged**